### PR TITLE
OPENEUROPA-3160: Update Timeline field type.

### PR DESCRIPTION
**Pull request by 22Alexandra**

### CODE REVIEW

File: modules/oe_content_timeline_field/src/Plugin/Field/FieldWidget/TimelineFieldWidget.php [hunk 1]
```diff
@@ -67,4 +70,51 @@ public function massageFormValues(array $values, array $form, FormStateInterface
...
+   * {@inheritdoc}
+   *
+   * Override the parameters to use the form element labels.
+   */
+  public function flagErrors(FieldItemListInterface $items, ConstraintViolationListInterface $violations, array $form, FormStateInterface $form_state) {
+    /** @var \Symfony\Component\Validator\ConstraintViolationInterface $violation */
+    foreach ($violations as $offset => $violation) {
+      $parameters = $violation->getParameters();
+      if (isset($parameters['%label'])) {
+        $parameters['%label'] = t('Label');
```
Comment (sergepavle) [inline modules/oe_content_timeline_field/src/Plugin/Field/FieldWidget/TimelineFieldWidget.php:96]:
```suggestion
        $parameters['%label'] = $this->t('Label');
```

File: modules/oe_content_timeline_field/tests/modules/oe_timeline_test_constraint/src/Plugin/Validation/Constraint/TestConstraintValidator.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    $parameters = [
+      '%label' => 'label',
+      '%title' => 'title',
+      '%body' => 'body',
+    ];
+    foreach ($value as $delta => $item) {
+      /** @var \Drupal\oe_content_timeline_field\Plugin\Field\FieldType\TimelineFieldItem $item */
+      $value = $item->getValue();
+      if ($value['label'] === '' || $value['label'] === NULL) {
```
Comment (sergepavle) [inline modules/oe_content_timeline_field/tests/modules/oe_timeline_test_constraint/src/Plugin/Validation/Constraint/TestConstraintValidator.php:31]: Could be single 'If' statement and better if first will be checking of body value.

File: modules/oe_content_timeline_field/tests/modules/oe_timeline_test_constraint/src/Plugin/Validation/Constraint/TestConstraintValidator.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      return;
+    }
+
+    $parameters = [
+      '%label' => 'label',
+      '%title' => 'title',
+      '%body' => 'body',
+    ];
+    foreach ($value as $delta => $item) {
+      /** @var \Drupal\oe_content_timeline_field\Plugin\Field\FieldType\TimelineFieldItem $item */
```
Comment (sergepavle) [inline modules/oe_content_timeline_field/tests/modules/oe_timeline_test_constraint/src/Plugin/Validation/Constraint/TestConstraintValidator.php:29]: Leftover

File: modules/oe_content_timeline_field/tests/modules/oe_timeline_test_constraint/oe_timeline_test_constraint.module [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+declare(strict_types = 1);
+
+use Drupal\Core\Entity\EntityTypeInterface;
+
+/**
+ * Implements hook_entity_bundle_field_info_alter().
+ */
+function oe_timeline_test_constraint_entity_bundle_field_info_alter(&$fields, EntityTypeInterface $entity_type, $bundle) {
+  // Add constraint to the paragraphs field of the Timeline paragraph.
```
Comment (brummbar) [inline modules/oe_content_timeline_field/tests/modules/oe_timeline_test_constraint/oe_timeline_test_constraint.module:16]: Uhm doubt.

File: modules/oe_content_timeline_field/tests/modules/oe_timeline_test_constraint/oe_timeline_test_constraint.info.yml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+name: OpenEuropa Timeline Test Constraint
+type: module
+description: Test the timeline widget when there are constraints set.
```
Comment (brummbar) [inline modules/oe_content_timeline_field/tests/modules/oe_timeline_test_constraint/oe_timeline_test_constraint.info.yml:3]: Provides a test constraint applied to a node timeline field.

File: modules/oe_content_timeline_field/tests/Functional/TimelineFieldValidationTest.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  public function testTimelineWidgetValidation() {
+    $user = $this->drupalCreateUser([], NULL, TRUE);
+    $this->drupalLogin($user);
+
+    $values = [
+      'title[0][value]' => 'My page',
+      'timeline[0][body][value]' => 'Body',
+    ];
+    $this->drupalPostForm('/node/add/page', $values, 'Save');
+    $this->assertSession()->pageTextContains('Label and Title fields cannot be empty when Content is specified.');
```
Comment (brummbar) [inline modules/oe_content_timeline_field/tests/Functional/TimelineFieldValidationTest.php:75]: We need to test that the correct elements are flagged with an error too.

File: modules/oe_content_timeline_field/tests/modules/oe_timeline_test_constraint/src/Plugin/Validation/Constraint/TestConstraintValidator.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+      $value = $item->getValue();
+      if (($value['body'] !== '' || $value['body'] !== NULL) &&
```
[the remark this answers is not in the document]
Comment (22Alexandra) [reply modules/oe_content_timeline_field/tests/modules/oe_timeline_test_constraint/src/Plugin/Validation/Constraint/TestConstraintValidator.php:31]: I would leave it like this to test the parameters replace too.

[the remark this answers is not in the document]
Comment (brummbar) [reply modules/oe_content_timeline_field/tests/modules/oe_timeline_test_constraint/src/Plugin/Validation/Constraint/TestConstraintValidator.php:31]: Yes the parameter replace is fine, we can just put a generic message that involves all parameters.
But with the current code, we use only $delta as path. There is no test coverage for a single sub-element.

File: modules/oe_content_timeline_field/src/Plugin/Field/FieldWidget/TimelineFieldWidget.php [hunk 8]
```diff
@@ -67,4 +71,51 @@ public function massageFormValues(array $values, array $form, FormStateInterface
...
+   * {@inheritdoc}
+   *
+   * Override the parameters to use the form element labels.
+   */
+  public function flagErrors(FieldItemListInterface $items, ConstraintViolationListInterface $violations, array $form, FormStateInterface $form_state) {
+    /** @var \Symfony\Component\Validator\ConstraintViolationInterface $violation */
+    foreach ($violations as $offset => $violation) {
+      $initial_parameters = $violation->getParameters();
+      $new_parameters = $initial_parameters;
+      if (isset($initial_parameters['%label']) && $initial_parameters['%label'] !== 'Label') {
```
Comment (brummbar) [inline modules/oe_content_timeline_field/src/Plugin/Field/FieldWidget/TimelineFieldWidget.php:94]: It's good, but there's no need to check if the value is the same. Let's just replace it and then check if any diffs are available.

File: modules/oe_content_timeline_field/src/Plugin/Field/FieldWidget/TimelineFieldWidget.php [hunk 9]
```diff
@@ -67,4 +70,51 @@ public function massageFormValues(array $values, array $form, FormStateInterface
...
+        $parameters['%label'] = $this->t('Label');
+      }
+      if (isset($parameters['%title'])) {
+        $parameters['%title'] = $this->t('Title');
+      }
+      if (isset($parameters['%body'])) {
+        $parameters['%body'] = $this->t('Content');
+      }
+      $violations->set($offset, new ConstraintViolation(
+        // @codingStandardsIgnoreStart
```
[the remark this answers is not in the document]
Comment (sergepavle) [reply modules/oe_content_timeline_field/src/Plugin/Field/FieldWidget/TimelineFieldWidget.php]: But why we are doing removing of the closing part?

File: modules/oe_content_timeline_field/src/Plugin/Field/FieldWidget/TimelineFieldWidget.php [hunk 10]
```diff
@@ -67,4 +71,55 @@ public function massageFormValues(array $values, array $form, FormStateInterface
...
+  /**
+   * {@inheritdoc}
+   *
+   * Override the parameters to use the form element labels.
+   */
+  public function flagErrors(FieldItemListInterface $items, ConstraintViolationListInterface $violations, array $form, FormStateInterface $form_state) {
+    /** @var \Symfony\Component\Validator\ConstraintViolationInterface $violation */
+    foreach ($violations as $offset => $violation) {
+      $initial_parameters = $violation->getParameters();
+      $parameters = $initial_parameters;
```
Comment (sergepavle) [inline modules/oe_content_timeline_field/src/Plugin/Field/FieldWidget/TimelineFieldWidget.php:93]:
```suggestion
       $parameters = $initial_parameters = $violation->getParameters();
```
could be like that. But it is optional comment.

**Pull request merged**